R/export_tibble.R

```diff
@@ -51,53 +52,93 @@ check_var_type <- function(GADSdat) {
+  if(identical(labeled, "yes") & all(is.na(label_df$value))) attr_list[["class"]] <- c("haven_labelled_spss")
+  any_miss <- !is.null(attr_list$na_values) || !is.null(attr_list$na_range)
+  if(identical(attr_list[["class"]], "haven_labelled") && any_miss) attr_list[["class"]] <- c("haven_labelled_spss", "haven_labelled")
 
   # experimental class modification (due to haven 2.3.0 classes got modified)
   #stopifnot(varClass %in% c("numeric", "character"))
-  #out[["class"]] <- c(out[["class"]], "vctrs_vctr", ifelse(varClass == "numeric", yes = "double", no = "character"))
+  #attr_list[["class"]] <- c(attr_list[["class"]], "vctrs_vctr", ifelse(varClass == "numeric", yes = "double", no = "character"))
 
   # value labels, if any
   value_label_df <- label_df[!is.na(label_df$value), ]
   if(nrow(value_label_df) > 0) {
-    out[["labels"]] <- value_label_df[, "value"]
-    names(out[["labels"]]) <- value_label_df[, "valLabel"]
+    attr_list[["labels"]] <- value_label_df[, "value"]
+    names(attr_list[["labels"]]) <- value_label_df[, "valLabel"]
   }
 
   # value labels and missing codes need to have the same class as the variable format
-  #if(all(label_df$varName == "groupVar")) browser()
-  if((length(out[["format.spss"]]) > 0 && grepl("^A", unique(out[["format.spss"]]))) || identical(varClass, "character")) {
-    #if((length(out[["format.spss"]]) > 0 && grepl("^A", unique(out[["format.spss"]])))) {
-    out[["na_values"]] <- as.character(out[["na_values"]])
+  if((length(attr_list[["format.spss"]]) > 0 &&
+      grepl("^A", unique(attr_list[["format.spss"]]))) ||
+     identical(varClass, "character")) {
+    attr_list[["na_values"]] <- as.character(attr_list[["na_values"]])
     if(nrow(value_label_df) > 0) {
-      out[["labels"]] <- as.character(out[["labels"]])
-      names(out[["labels"]]) <- value_label_df[, "valLabel"]
+      attr_list[["labels"]] <- as.character(attr_list[["labels"]])
+      names(attr_list[["labels"]]) <- value_label_df[, "valLabel"]
     }
-    #if(identical(labeled, "yes")) out[["class"]] <- c("haven_labelled")
+    #if(identical(labeled, "yes")) attr_list[["class"]] <- c("haven_labelled")
   }
 
-  #if(any(label_df$varName == "TESTUNG")) browser()
-  out
+  attr_list
 }
 
+# missing tag conversion to suit SPSS style (maximum 3 discrete missing tags or range of missing tags)
+add_miss_tags <- function(varName, attr_list, label_df, raw_dat) {
+  miss_values <- label_df[which(label_df$missings == "miss"), "value"]
+
+  if(length(miss_values) > 0 && length(miss_values) <= 3)  {
+    attr_list[["na_values"]] <- miss_values
+  }
+  if(length(miss_values) > 3) {
+    full_range <- range(miss_values)
+    # check variable class
+    if(is.character(raw_dat)) {
+      stop("Missing tag conversion for variable '", varName,
+           "' to SPSS conventions is not possible, as too many missings are declared for a character variable (maximum 3). Adjust missing tags to export to tibble or write to SPSS.")
+    }

```

Comment:
Nit: Maybe, this could be restructured to (to overview the condition tree faster):

```r
  if (length(miss_values) > 0) {
    if (length(miss_values) <= 3) {
      attr_list[["na_values"]] <- miss_values
    } else {
      full_range <- range(miss_values)
      # check variable class
      if (is.character(raw_dat)) {
        stop("Missing tag conversion for variable '", varName,
             "' to SPSS conventions is not possible, as too many missings are declared for a character variable (maximum 3). Adjust missing tags to export to tibble or write to SPSS.")
      ... # Rest of the code block
      }
    }
  }
```

Reply:
As mentioned above, I have a strong perference of avoiding `else`. However, I do agree that it is more readable in this case! Modified as suggested!